### PR TITLE
feat(restricted_glob): negated globs

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

In the [import revamping RFC](https://github.com/biomejs/biome/discussions/3015), I proposed that each group be represented by a special identifier or a glob.
For example in the following list we have two groups:

```json5
{
  groups: [
    "@myCompany/**"
    "@*/**"
  ]
}
```

This will place all imports starting with `@myCompany/` before all other scoped imports.
Unfortunately, as noted by a [comment in the RFC](https://github.com/biomejs/biome/discussions/3015#discussioncomment-9586004), this doesn't allow doing the reverse, i.e. placing first all coped imports (except the ones starting with `@myCompany/`).
At the time, I thought about a [recursive approach](https://github.com/biomejs/biome/discussions/3015#discussioncomment-9639600) to solve this limitation.
However, this looks complicated and might even produce unwanted or confusing results for users.
I think we should keep a simple approach where we stop as soon as a group matches.
So to solve the previous limitation, I think we could allow a group to be defined by a list of globs.
We could allow the inclusion of negated globs in this list, which would act as an exception (similar to what `.gitignore` does).
Then we could write:

```json5
{
  groups: [
    ["@*/**", "!@myCompany/**"],
    "@myCompany/**"
  ]
}
```

This PR adds negated globs to restricted globs and a method `matches_with_exceptions` to match against a list of globs where negated globs are treated as exceptions (`.gitignore`-like negated globs).

## Test Plan

I added some unit tests.

### CODE REVIEW

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 1]
```diff
@@ -98,12 +125,40 @@ impl schemars::JsonSchema for RestrictedGlob {
...
+    }
+
+    /// Match against a list of globs where negated globs are handled as exceptions.
+    ///
+    /// This mimics the behavior of `.gitignore`.
+    /// For example, given the list `["**/*", "!a*", "aa", "!a"]`, we have:
+    /// - The file `a` doesn't match
+    /// - The file `aa` matches
+    /// - The file `aaa` doesn't match
+    /// - The file `b` matches
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:148]: This is **hard** to understand, really.

Here's a few pointers to make it simpler:
- don't mention gitignore, because some people might not know the feature you mentioned 
- the examples should explain **why they match/not match**. For example, I don't... And still can't 
- add doc tests, if you really want to have examples 
- you deferred the "exception" concept to gitignore, but didn't provide any link to it 

I suggest you explain the feature yourself, and avoid deleting to other resources

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 2]
```diff
@@ -98,12 +125,40 @@ impl schemars::JsonSchema for RestrictedGlob {
...
+    pub fn matches_with_exceptions<'b, I>(&self, globs: I) -> bool
+    where
+        I: IntoIterator<Item = &'b RestrictedGlob>,
+        I::IntoIter: DoubleEndedIterator,
+    {
+        // Iterate in reverse order to avoid unnecessary tests.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:154]: I suggest rewording the comment. What tests? At the beginning I thought you were talking about **real** tests...

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 3]
```diff
@@ -7,27 +7,46 @@ use biome_rowan::{TextRange, TextSize};
...
+        self.is_negated
+    }
+
     /// Tests whether the given path matches this pattern or not.
     pub fn is_match(&self, path: impl AsRef<std::path::Path>) -> bool {
-        self.0.is_match(path)
+        self.is_raw_match(path) != self.is_negated
+    }
+
+    /// Tests whether the given path matches this pattern or not ignoring the negation.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:30]:
```suggestion
    /// Tests whether the given path matches this pattern, ignoring the negation
```

When you use the word "whether", the "or not" is always implied by the word itself

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 4]
```diff
@@ -7,27 +7,46 @@ use biome_rowan::{TextRange, TextSize};
...
+    is_negated: bool,
+    glob: globset::GlobMatcher,
+}
 impl RestrictedGlob {
+    /// Returns `true` if this glob is negated.
+    pub fn is_negated(&self) -> bool {
+        self.is_negated
+    }
+
     /// Tests whether the given path matches this pattern or not.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:25]:
```suggestion
    /// Tests whether the given path matches this pattern.
```

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 5]
```diff
@@ -7,27 +7,46 @@ use biome_rowan::{TextRange, TextSize};
...
-        self.0.is_match(path)
+        self.is_raw_match(path) != self.is_negated
+    }
+
+    /// Tests whether the given path matches this pattern or not ignoring the negation.
+    fn is_raw_match(&self, path: impl AsRef<std::path::Path>) -> bool {
+        self.glob.is_match(path)
     }
 
     /// Tests whether the given path matches this pattern or not.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:35]:
```suggestion
    /// Tests whether the given path matches this pattern.
```

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 6]
```diff
@@ -7,27 +7,46 @@ use biome_rowan::{TextRange, TextSize};
...
     pub fn is_match_candidate(&self, path: &CandidatePath<'_>) -> bool {
-        self.0.is_match_candidate(&path.0)
+        self.is_raw_match_candidate(path) != self.is_negated
+    }
+
+    /// Tests whether the given path matches this pattern or not ignoring the negation.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:40]:
```suggestion
    /// Tests whether the given path matches this pattern, ignoring the negation.
```

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 7]
```diff
@@ -98,12 +125,40 @@ impl schemars::JsonSchema for RestrictedGlob {
...
 /// path against multiple globs or sets of globs.
+#[derive(Debug, Clone)]
 pub struct CandidatePath<'a>(globset::Candidate<'a>);
 impl<'a> CandidatePath<'a> {
     /// Create a new candidate for matching from the given path.
     pub fn new(path: &'a impl AsRef<std::path::Path>) -> Self {
         Self(globset::Candidate::new(path))
     }
+
+    /// Tests whether the current path matches `glob` or not.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:136]:
```suggestion
    /// Tests whether the current path matches `glob`.
```

File: crates/biome_js_analyze/src/utils/restricted_glob.rs [hunk 8]
```diff
@@ -98,12 +135,59 @@ impl schemars::JsonSchema for RestrictedGlob {
...
+    /// assert!(!CandidatePath::new(&"abc").matches_with_exceptions(globs));
+    /// ```
+    ///
+    /// - `b` matches `*` and is not excluded by the exception `!a*`.
+    ///   Thus, `b` matches the list of globs.
+    /// - `abc` matches the first glob `*`, however it is excluded by the exception `!a*`.
+    ///   Thus `abc` doesn't match the list of globs.
+    /// - `a` matches the first glob `*` and is excluded by the exception `!a*`.
+    ///   However, it is included again by the last glob `a`.
+    ///   Thus `a` matches the list of globs.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/utils/restricted_glob.rs:176]: Thank you for providing a very good explanation. 

Damn, these globs can create a lot confusion, based on this explanation. I don't know how to feel about them, because they have too much power. 

For example, `["a*", "!a", "a*", "!a*", "a"]`, which is an absurd edge case, means that the input glob `a` will match the last `a`. Does that mean that we need to check **all** globs? Or is that why we use `rev()`? 

Then I hope we actually explain that in our docs, and tell the users that the last glob is the one that wins.

Comment (Conaclos) [reply crates/biome_js_analyze/src/utils/restricted_glob.rs:176]: There are two ways of implementing it:

- From left to right (aka in-order)
  1. Looks for the first non-negated glob that matches
  2. From this first matching glob, looks for the first negated glob that doesn't match
  3. repeat
  4. If we end in step (1) then we have a match, otherwise we have no match
- From right to left (aka in reverse order)
  Looks for the first non-negated glob that matches or the first negated glob that matches
  If it is a non-negated glob, we have a match, otherwise we have no match

I choose the reverse order because it looks simpler to me and could avoid some glob matching.

You edge case that you present has the intended behavior. `.gitignore` has the same behavior.

Do you think I should add this edge case in the docs?

Comment (ematipico) [reply crates/biome_js_analyze/src/utils/restricted_glob.rs:176]: Thank you for the explanation. 

> Do you think I should add this edge case in the docs?

Yes, but in the user docs when we teach them how to use the options :)
